// @biit-solutions/authorization-services/esm2020/lib/services/auth.service.mjs
import { User } from "../models/user";
export class AuthService {
    constructor(rootPath, httpClient) {
        this.rootPath = rootPath;
        this.httpClient = httpClient;
    }
    getAll() {
        return this.httpClient.get(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/register`);
    }
    add(user) {
        return this.httpClient.post(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/register`, user);
    }
    update(request) {
        return this.httpClient.patch(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/register`, request);
    }
    login(request) {
        return this.httpClient.post(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/public/login`, request, { observe: 'response' });
    }
    changePassword(request) {
        return this.httpClient.post(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/password`, request);
    }
    getRoles() {
        return this.httpClient.get(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/roles`);
    }
    deleteByUserName(username) {
        return this.httpClient.delete(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/register/${username}`);
    }
    cancelAutoRenew() {
        clearTimeout(this.timeoutId);
    }
    loginByToken(token) {
        return this.httpClient.get(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/public/login/tokens/${token}`, { observe: 'response' });
    }
    autoRenewToken(token, expiration, tokenRenewListener, tolerance = AuthService.TOLERANCE) {
        if (this.timeoutId != null) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
        const expirationDate = new Date(expiration);
        const now = new Date();
        const expirationTime = expirationDate.getTime() - AuthService.TOLERANCE - now.getTime();
        this.setTimeoutRenew(token, expirationTime, tokenRenewListener, tolerance);
        console.log(`Next token renew on: ${expirationDate}`);
    }
    setTimeoutRenew(token, timeout, tokenRenewListener, tolerance) {
        if (isNaN(timeout)) {
            console.error(`Timeout should be a number and received '${timeout}'`);
            return;
        }
        if (timeout < AuthService.MIN_WAIT_TIME) {
            timeout = AuthService.MIN_WAIT_TIME;
        }
        this.timeoutId = setTimeout(() => {
            this.renew(token).subscribe({
                next: (res) => {
                    if (!res) {
                        console.error('Server returned invalid response');
                        return;
                    }
                    const authToken = res.headers.get('authorization');
                    const expiration = Number(res.headers.get('expires'));
                    const user = User.clone(res.body);
                    if (!authToken || !expiration) {
                        throw new Error('Server returned invalid response');
                    }
                    if (isNaN(expiration)) {
                        throw new Error('Server returned invalid expiration time');
                    }
                    console.log(`Next token renew on: ${new Date(expiration)}`);
                    tokenRenewListener.onTokenReceived(authToken, expiration, user);
                    const newTimeOut = expiration - new Date().getTime() - tolerance;
                    this.setTimeoutRenew(authToken, newTimeOut, tokenRenewListener, tolerance);
                },
                error: (err) => {
                    tokenRenewListener.onException(err);
                    clearTimeout(this.timeoutId);
                }
            });
        }, timeout);
    }
    renew(token) {
        return token ?
            this.httpClient.get(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/jwt/renew`, { headers: { 'Authorization': 'Bearer ' + token }, observe: 'response' })
            : this.httpClient.get(`${this.rootPath.getRootPath()}${AuthService.ROOT_PATH}/jwt/renew`, { observe: 'response' });
    }
}
AuthService.ROOT_PATH = '/auth';
AuthService.MIN_WAIT_TIME = 1000;
AuthService.TOLERANCE = 1000 * 60 * 5; // 5 minutes
