// @biit-solutions/authorization-services/esm2020/lib/services/session.service.mjs
import { Constants } from "../constants";
import { User } from "../models/user";
export class SessionService {
    constructor(authService, context = '') {
        this.authService = authService;
        this.context = context;
        this.loggedIn = false;
        const authToken = localStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
        const expires = +localStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
        let user = localStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`) ? User.clone(JSON.parse(localStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`))) : undefined;
        if (!user) {
            user = sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`) ? User.clone(JSON.parse(sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`))) : undefined;
        }
        this.user = user;
        if (!expires || isNaN(expires) || expires < new Date().getTime()) {
            localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
            localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
            localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`);
        }
        if (authToken) {
            sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`, authToken);
            sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`, JSON.stringify(user));
            this.store = true;
            if (expires && !isNaN(expires)) {
                sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`, expires.toString());
            }
            this.loggedIn = true;
        }
        if ((authToken && expires)
            || (sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`) && sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`))) {
            this.setAutoRenew(authToken ? authToken : sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`), expires ? expires : +sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`));
        }
    }
    onTokenReceived(token, expiration, user) {
        this.setToken(token, expiration);
        this.setUser(user);
        console.log(`Token ${this.context ? 'for' + this.context : ''} has been renewed successfully.`, token);
    }
    onException(error) {
        console.error(`There was an exception while renewing the token ${this.context ? 'for' + this.context : ''}.`);
        this.clearToken();
    }
    setAutoRenew(token, expires) {
        // ADD HERE ALL SERVICES NEED TO BE CALLED TO RENEW TOKEN
        this.authService.autoRenewToken(token, expires, this);
    }
    clearToken() {
        sessionStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
        sessionStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
        sessionStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`);
        this.store = undefined;
        localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
        localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
        localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`);
        this.loggedIn = false;
        this.user = undefined;
    }
    setToken(token, expires, enableStore = undefined, autoRenew = false) {
        if (!token || !expires) {
            this.clearToken();
            return;
        }
        sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`, token);
        sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`, expires.toString());
        if (enableStore !== undefined) {
            this.store = enableStore;
            if (!enableStore) {
                localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
                localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
            }
        }
        if (this.store) {
            localStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`, token);
            localStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`, expires.toString());
        }
        if (autoRenew) {
            this.setAutoRenew(token, expires);
        }
        this.loggedIn = true;
    }
    getToken() {
        return sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_TOKEN}`);
    }
    isTokenExpired() {
        const expired = !sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`) ||
            new Date().getTime() > +sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`) || !this.getToken();
        if (!expired) {
            this.loggedIn = true;
        }
        return expired;
    }
    getExpirationDate() {
        const sessionExpiration = sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`);
        if (!isNaN(+sessionExpiration)) {
            return new Date(+sessionStorage.getItem(`${this.context}.${Constants.SESSION_STORAGE.AUTH_EXPIRATION}`));
        }
        return null;
    }
    get isLoggedIn() {
        return this.loggedIn;
    }
    setUser(user, enableStore = undefined) {
        sessionStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`, JSON.stringify(user));
        if (enableStore !== undefined) {
            if (!enableStore) {
                localStorage.removeItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`);
            }
        }
        if (this.store) {
            localStorage.setItem(`${this.context}.${Constants.SESSION_STORAGE.USER}`, JSON.stringify(user));
        }
        this.user = user;
    }
    getUser() {
        return this.user;
    }
}
