// @biit-solutions/authorization-services/esm2020/lib/models/app-role.mjs
export var AppRole;
(function (AppRole) {
    AppRole["APPOINTMENTCENTER_ADMIN"] = "APPOINTMENTCENTER_ADMIN";
    AppRole["APPOINTMENTCENTER_MANAGER"] = "APPOINTMENTCENTER_MANAGER";
    AppRole["APPOINTMENTCENTER_USER"] = "APPOINTMENTCENTER_USER";
    AppRole["BASEFORMDROOLSENGINE_ADMIN"] = "BASEFORMDROOLSENGINE_ADMIN";
    AppRole["BASEFORMDROOLSENGINE_USER"] = "BASEFORMDROOLSENGINE_USER";
    AppRole["FACTMANAGER_ADMIN"] = "FACTMANAGER_ADMIN";
    AppRole["FACTSDASHBOARD_CEO"] = "FACTSDASHBOARD_CEO";
    AppRole["FACTMANAGER_USER"] = "FACTMANAGER_USER";
    AppRole["INFOGRAPHICENGINE_ADMIN"] = "INFOGRAPHICENGINE_ADMIN";
    AppRole["INFOGRAPHICENGINE_USER"] = "INFOGRAPHICENGINE_USER";
    AppRole["KAFKAPROXY_ADMIN"] = "KAFKAPROXY_ADMIN";
    AppRole["KAFKAPROXY_USER"] = "KAFKAPROXY_USER";
    AppRole["KNOWLEDGESYSTEM_ADMIN"] = "KNOWLEDGESYSTEM_ADMIN";
    AppRole["KNOWLEDGESYSTEM_USER"] = "KNOWLEDGESYSTEM_USER";
    AppRole["PROFILEMATCHER_ADMIN"] = "PROFILEMATCHER_ADMIN";
    AppRole["PROFILEMATCHER_MANAGER"] = "PROFILEMATCHER_MANAGER";
    AppRole["PROFILEMATCHER_USER"] = "PROFILEMATCHER_USER";
    AppRole["TASKJUNCTION_ADMIN"] = "TASKJUNCTION_ADMIN";
    AppRole["TASKJUNCTION_USER"] = "TASKJUNCTION_USER";
    AppRole["USERMANAGERSYSTEM_ADMIN"] = "USERMANAGERSYSTEM_ADMIN";
    AppRole["USERMANAGERSYSTEM_EDITOR"] = "USERMANAGERSYSTEM_EDITOR";
    AppRole["USERMANAGERSYSTEM_ORGANIZATION_ADMIN"] = "USERMANAGERSYSTEM_ORGANIZATION_ADMIN";
    AppRole["USERMANAGERSYSTEM_USER"] = "USERMANAGERSYSTEM_USER";
    AppRole["PROFILEMATCHER_ORGANIZATION_ADMIN"] = "PROFILEMATCHER_ORGANIZATION_ADMIN";
})(AppRole || (AppRole = {}));
