// @biit-solutions/authorization-services/esm2020/lib/models/role.mjs
export var Role;
(function (Role) {
    Role["APPOINTMENTCENTER_ADMIN"] = "APPOINTMENTCENTER_ADMIN";
    Role["APPOINTMENTCENTER_EDITOR"] = "APPOINTMENTCENTER_EDITOR";
    Role["APPOINTMENTCENTER_VIEWER"] = "APPOINTMENTCENTER_VIEWER";
    Role["BASEFORMDROOLSENGINE_ADMIN"] = "BASEFORMDROOLSENGINE_ADMIN";
    Role["BASEFORMDROOLSENGINE_EDITOR"] = "BASEFORMDROOLSENGINE_EDITOR";
    Role["BASEFORMDROOLSENGINE_VIEWER"] = "BASEFORMDROOLSENGINE_VIEWER";
    Role["INFOGRAPHICENGINE_ADMIN"] = "INFOGRAPHICENGINE_ADMIN";
    Role["INFOGRAPHICENGINE_EDITOR"] = "INFOGRAPHICENGINE_EDITOR";
    Role["INFOGRAPHICENGINE_PRACTITIONER"] = "INFOGRAPHICENGINE_PRACTITIONER";
    Role["INFOGRAPHICENGINE_VIEWER"] = "INFOGRAPHICENGINE_VIEWER";
    Role["KAFKAPROXY_ADMIN"] = "KAFKAPROXY_ADMIN";
    Role["KAFKAPROXY_EDITOR"] = "KAFKAPROXY_EDITOR";
    Role["KAFKAPROXY_VIEWER"] = "KAFKAPROXY_VIEWER";
    Role["KNOWLEDGESYSTEM_ADMIN"] = "KNOWLEDGESYSTEM_ADMIN";
    Role["KNOWLEDGESYSTEM_EDITOR"] = "KNOWLEDGESYSTEM_EDITOR";
    Role["KNOWLEDGESYSTEM_VIEWER"] = "KNOWLEDGESYSTEM_VIEWER";
    Role["FACTMANAGER_ADMIN"] = "FACTMANAGER_ADMIN";
    Role["FACTMANAGER_EDITOR"] = "FACTMANAGER_EDITOR";
    Role["FACTMANAGER_VIEWER"] = "FACTMANAGER_VIEWER";
    Role["FACTSDASHBOARD_ADMIN"] = "FACTSDASHBOARD_ADMIN";
    Role["FACTSDASHBOARD_VIEWER"] = "FACTSDASHBOARD_VIEWER";
    Role["FACTSDASHBOARD_CEO"] = "FACTSDASHBOARD_CEO";
    Role["FACTSDASHBOARD_TEAMLEADER"] = "FACTSDASHBOARD_TEAMLEADER";
    Role["FACTSDASHBOARD_EMPLOYEE"] = "FACTSDASHBOARD_EMPLOYEE";
    Role["FACTSDASHBOARD_PRACTITIONER"] = "FACTSDASHBOARD_PRACTITIONER";
    Role["FACTSDASHBOARD_NCA"] = "FACTSDASHBOARD_NCA";
    Role["FACTSDASHBOARD_CADT"] = "FACTSDASHBOARD_CADT";
    Role["FACTSDASHBOARD_CREDIBILITY"] = "FACTSDASHBOARD_CREDIBILITY";
    Role["FACTSDASHBOARD_CUSTOMERLIST"] = "FACTSDASHBOARD_CUSTOMER LIST";
    Role["FACTSDASHBOARD_HAPPINESSATWORK"] = "FACTSDASHBOARD_HAPPINESS AT WORK";
    Role["FACTSDASHBOARD_FRUSTRATIONONTEAMWORKING"] = "FACTSDASHBOARD_FRUSTRATION ON TEAMWORKING";
    Role["FACTSDASHBOARD_ORGANIZATIONLIST"] = "FACTSDASHBOARD_ORGANIZATION LIST";
    Role["FACTSDASHBOARD_XLSACCESS"] = "FACTSDASHBOARD_XLS ACCESS";
    Role["PROFILEMATCHER_ADMIN"] = "PROFILEMATCHER_ADMIN";
    Role["PROFILEMATCHER_EDITOR"] = "PROFILEMATCHER_EDITOR";
    Role["PROFILEMATCHER_VIEWER"] = "PROFILEMATCHER_VIEWER";
    Role["TASKJUNCTION_ADMIN"] = "TASKJUNCTION_ADMIN";
    Role["TASKJUNCTION_EDITOR"] = "TASKJUNCTION_EDITOR";
    Role["TASKJUNCTION_VIEWER"] = "TASKJUNCTION_VIEWER";
    Role["USERMANAGERSYSTEM_ADMIN"] = "USERMANAGERSYSTEM_ADMIN";
    Role["USERMANAGERSYSTEM_EDITOR"] = "USERMANAGERSYSTEM_EDITOR";
    Role["USERMANAGERSYSTEM_VIEWER"] = "USERMANAGERSYSTEM_VIEWER";
    Role["USERMANAGERSYSTEM_ORGANIZATION_ADMIN"] = "USERMANAGERSYSTEM_ORGANIZATION_ADMIN";
    Role["FACTSDASHBOARD_ORGANIZATION_ADMIN"] = "FACTSDASHBOARD_ORGANIZATION_ADMIN";
    Role["PROFILEMATCHER_ORGANIZATION_ADMIN"] = "PROFILEMATCHER_ORGANIZATION_ADMIN";
})(Role || (Role = {}));
